### ISSUE TITLE
Support GET request with query parameters from map

### ISSUE THREAD

**Issue by nmiyake** [Netflix/feign#330]
Currently, Feign does not provide native support for making a GET request with dynamic query parameters and values using a map. It would be great if there was a mechanism for providing a Map or Multimap of Strings whose contents would be used as the query parameters for a GET call.

I have a work-around using encoders as a PR in a different repository at https://github.com/palantir/http-remoting/pull/53, but wanted to open this issue to see if it would be possible to implement this directly in Feign.

**Comment by nmiyake** [comment]
It looks like this has been requested previously as https://github.com/Netflix/feign/issues/216, and although there are work-arounds it would be great to have better native support for this.

For reference, this is supported in Retrofit using a `@QueryMap` annotation (https://github.com/square/retrofit/issues/293).

**Comment by adriancole** [comment]
I'm open to proposals on this. I like that we currently don't have anything
except a Param (and know a custom expander is less than ideal for this).
I'm wondering if we can support this unambiguously (or at least mostly
unambiguous) without making new annotations?

For example, `@Param` on a type that is a map, which is a query parameter
is a narrow definition.. I wonder if we can get by with just defaulting
this circumstance to a query exploder?

**Comment by nmiyake** [comment]
Happy to explore an approach without extra annotations, but would likely need some guidance as to how to approach it.

Let's start with a method like the following:

```
@GET
@Path("/getUsingParamMap")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
Map<String, String> getUsingParamMap(@Param(value = "queryParams") Map<String, String> queryParams);
```

Currently, calling this will fail with an exception along the lines of `method GET must not have a request body.` (using `OkHttp`)

It looks like the method's body and query parameters are currently created set in `feign.ReflectiveFeign.BuildTemplateByResolvingArgs.create(Object[])`.

It should be simple enough to add logic here along the lines of "if method is a `GET` method with a `@Param` that's a `Map`, interpret as query parameters instead of a body" (as long as annotations etc. can be accessed).

Is that what you're suggesting, or is the query exploder something separate? Happy to give something a try if I get a general idea of what to do.

**Comment by adriancole** [comment]
thanks for the feedback.. Firstly I wasn't really thinking about this for jaxrs, rather the default contract.

I suspect to make this work, we'd need `MethodMetadata.queryMapIndex`, which stores where the special query map is. We'd definitely want to address this situation during contract parse

option 1: no new annotation.. just look at where it is.. (ex. a variable directly after '?' or '&')
```java
    @RequestLine("GET /domains/{domainId}/records?name={name}&{extra}")
    Response recordsByNameAndType(@Param("domainId") int id, @Param("name") String nameFilter,
                                  @Param("extra") Map<String, String> options);
```

option 2: new annotation which splats on the end
```java
    @RequestLine("GET /domains/{domainId}/records?name={name}")
    Response recordsByNameAndType(@Param("domainId") int id, @Param("name") String nameFilter,
                                  @QueryMap Map<String, String> options);
```

I'm mildly partial towards the former. cc @beku8

**Comment by adriancole** [comment]
if going with the former, what would happen is that MethodMetadata.queryMapIndex would trip a non-default Param.Expander, which would splat out the map in-place. To be portable, you'd only do that when the user hasn't already defined their own Param.Expander

**Comment by adriancole** [comment]
sheesh.. I forgot we already have MethodMetadata.indexToExpanderClass! this should actually be very easy to support since you don't need to change anything except DefaultContract for option 1: (ex. a variable directly after '?' or '&')

**Comment by nmiyake** [comment]
Thanks for the guidance, that sounds great! Fine with option 1 as well -- agree that it fits in nicely with existing infrastructure. Will give this a shot.

**Comment by nmiyake** [comment]
@adriancole , took a closer look at this, but don't think that the expanders will be enough on their own.

Let's take your example and assume that `recordsByNameAndType` is called with `Map<String, String> options` being `{"key1"="val1", "key2=val2"}`.

If some kind of "splatting" expander is defined, it will be triggered in `feign.ReflectiveFeign.BuildTemplateByResolvingArgs.create(Object[])`. In the example you cited, it would splat out the `@Param("extra")` map, so the `Map<String, Object> varBuilder` map be `{"extra"="key1=val1&key2=val2"}`.

The issue is that this map is passed directly to `feign.RequestTemplate.resolve(Map<String, ?> unencoded)`, and that method will process the contents as an unencoded String, so it will encode it first and thus strip out the concatenation.

Based on the way the current logic is structure, I don't think it's possible to fit this in using the existing structure exactly. The approach that I can imagine right now looks something like this:

* `Contract.processAnnotationsOnParameter` needs to recognize query map variable and not interpret it as a formParam
* `RequestTemplate` has an extra String field that stores the parameter name for a QueryMap parameter (`null` if no expanding map parameter is specified)
* When `RequestTemplate` process the query String in `pullAnyQueriesOutOfUrl`, if it recognizes a QueryMap parameter (a variable in a query String that comes directly after '?' or '&'), it stores that in the new field
* When `RequestTemplate` runs `replaceQueryValues`, if the new QueryMap param is non-null, its value is retrieved from the "unencoded" map. This value is the Map of query params to query param values. The method will process this map and add the entires to the `queries` map, encoding the values as necessary

This may not be the optimal approach, but I think that any approach will need to modify things beyond the current structure (because the current structure is pretty strict about how queries are constructed, even with expanders).

I have local changes that do the above and can push for reference if it would be helpful to see. However, wanted to check on what you thought of the general approach before I spend too much time touching up the code, writing tests etc.

**Comment by nmiyake** [comment]
Playing around some more, the step of "`Contract.processAnnotationsOnParameter` needs to recognize query map variable and not interpret it as a formParam" is actually pretty hard to do with a generic `@Param` annotation because there's not much data accessible at this point to be able to make the determination that we want this to be a query param map. Based on this, I think approach 2 (new annotation that splats on the end) may be cleaner to implement -- going to take a look at that.

**Comment by adriancole** [comment]
hi, there.. I think I get the general gist of the problem. I'm apprehensive
about you spending too much more time here.. you are probably right that
some internals aren't helping, but maybe we can special case somehow to
permit this pattern and not affect too much code. I'll have a look in case
I can identify a shortcut..

**Comment by adriancole** [comment]
wires crossed! feel free to explore.. I don't want to hold you back. My aim
is to be able to introduce this with least code and incompatibility. (ex,
don't want to break the contract interface (although breaking
DefaultContract is fine)

**Comment by nmiyake** [comment]
I wrote a simple example implementation of approach 2 -- viewable at PR #335 . Let me know what you think of the general approach -- if it looks promising, I can clean up the code, make the expected behavior more explicit and add more tests.

General approach:

* New `@QueryMap` annotation is introduced
* Index of all parameters with this annotation is stored in a new variable in `MethodMetadata`
* After the request template is constructed and processed normally, all of the `QueryMap` annotated parameters are processed in order. For any annotated parameter that is a `Map` (technically must be `Map<String, ?>`), the content of the map is added as query parameters
  * Values are added as their string representation unless the object implements `Iterable`, in which case each element's String representation is set as values on the key

This approach involves fairly little code and is pretty well-defined and clean from a conceptual standpoint -- it preserves back-compat, doesn't require editing any paths, plays nicely with previous annotations and explicitly defined query parameters, and leverages existing code/implementations. I think it's much cleaner than trying to mess with the `RequestTemplate` code to change param parsing/interpretation. Let me know what you think.

**Comment by adriancole** [comment]
nice work!

**Comment by nmiyake** [comment]
Fixed by #335 . Thanks for all the guidance on this!

**Comment by nmiyake** [comment]
@adriancole , what is the release process for Feign? Just wanted to get a general idea of when the next release with this fix will be, and couldn't find any documentation about what triggers releases (the Wiki has information about how to execute releases, but not on the cadence or release criteria). Thanks!

**Comment by adriancole** [comment]
release trigger is a tag, triggering the tag is a nag :P

**Comment by nmiyake** [comment]
@adriancole friendly nag for a tag :). Know that you were waiting on the InvocationHandlerFactory stuff, but just wanted to check in.

**Comment by adriancole** [comment]
8.15.0 is on the way! https://travis-ci.org/Netflix/feign/builds/113839780

**Comment by inad9300** [comment]
This does work with POJOs, only with maps :/

By "this" I mean: https://github.com/OpenFeign/feign/pull/335/files#diff-13815c678552a275075ab18b17c67784R616

**Comment by inad9300** [comment]
I have created a new issue for this problem: https://github.com/OpenFeign/feign/issues/852
